File: src/nric.py
from math import floor


class NRICValidator(object):
    """
    Validates a Singaporean FIN or NRIC to ensure that it's algorithmically correct
    
    >>> NRICValidator.is_nric_valid('S8944027J')
    True
    >>> NRICValidator.is_fin_valid('G6046409Q')
    True
    >>> NRICValidator.is_nric_valid('S814A299Z')
    False
    >>> NRICValidator.is_fin_valid('G6075119Q')
    False
    >>> NRICValidator.is_nric_valid('123')
    False
    >>> NRICValidator.is_fin_valid('123')
    False
    """
    multiples = [2, 7, 6, 5, 4, 3, 2]
    
    @classmethod
    def is_valid(cls, nric_or_fin):
        validator = cls()
        if validator.is_nric_valid(nric_or_fin):
            return True
        elif validator.is_fin_valid(nric_or_fin):
            return True
        else:
            return False

    @classmethod
    def is_nric_valid(cls, the_nric):
    
        if not the_nric:
            return False
    
        if len(the_nric) != 9:
            return False;
    
        first = the_nric[0]
        last = the_nric[-1];
    
        if first not in ['S','T']:
            return False
    
        try:
            numeric = int(the_nric[1:-1])
        except ValueError:
            return False
                
        if first == 'S':
            outputs = ['J', 'Z', 'I', 'H', 'G', 'F', 'E', 'D', 'C', 'B', 'A']
        else:
            outputs = ['G', 'F', 'E', 'D', 'C', 'B', 'A', 'J', 'Z', 'I', 'H']
    
        return cls.check_mod_11(last, numeric, outputs)
        
    @classmethod
    def is_fin_valid (cls, fin):
        
        if not fin:
            return False
    
        if len(fin) != 9:
            return False
    
        first = fin[0]
        last = fin[-1]
    
        if first not in ['F', 'G']:
            return False

        try:
            numeric = int(fin[1:-1])
        except ValueError:
            return False
    
        if first == 'F':
            outputs = ['X', 'W', 'U', 'T', 'R', 'Q', 'P', 'N', 'M', 'L', 'K']
        else:
            outputs = ['R', 'Q', 'P', 'N', 'M', 'L', 'K', 'X', 'W', 'U', 'T']
    
        return cls.check_mod_11(last, numeric, outputs)
        
    @classmethod
    def check_mod_11(cls, last, numeric, outputs):

        total = count = 0
 
        while numeric != 0:
            total += (numeric % 10) * self.multiples[len(self.multiples) - (1 + count)]
            
            count += 1
    
            numeric /= 10
            numeric = floor(numeric)
        
        return last == outputs[int(total % 11)]


if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: src/nric.go
/*
 * Compile and run with 6g nric.go && 6l nric.6 && ./6.out
*/

package main

import (
    "fmt"
    "strconv"
)

var multiples = []int { 2, 7, 6, 5, 4, 3, 2 }

func isNricValid (theNric string) bool {

        if len(theNric) != 9 {
                return false
        }

        var first int = int(theNric[0])
        var last  int = int(theNric[len(theNric)-1])
        var outputs []int

        if first != 'S' && first != 'T' {
            return false
        }

        numeric, err := strconv.Atoi(theNric[1:len(theNric)-1])

        if err != nil {
                return false
        }

        if first == 'S' {
            outputs = []int{ 'J', 'Z', 'I', 'H', 'G', 'F', 'E', 'D', 'C', 'B', 'A' }
        } else {
            outputs = []int{ 'G', 'F', 'E', 'D', 'C', 'B', 'A', 'J', 'Z', 'I', 'H' }
        }

        return checkMod11(last, numeric, outputs)

}

func isFinValid(theFin string) bool {

        if len(theFin) != 9 {
                return false
        }

        var first int = int(theFin[0])
        var last int = int(theFin[len(theFin)-1])
        var outputs []int

        if first != 'G' && first != 'F' {
            return false
        }

        numeric, err := strconv.Atoi(theFin[1:len(theFin)-1])

        if err != nil {
                return false
        }

        if first == 'F' {
            outputs = []int{ 'X', 'W', 'U', 'T', 'R', 'Q', 'P', 'N', 'M', 'L', 'K' }
        } else {
            outputs = []int{ 'R', 'Q', 'P', 'N', 'M', 'L', 'K', 'X', 'W', 'U', 'T' }
        }

        return checkMod11(last, numeric, outputs)

}

func checkMod11 (last int, numeric int, outputs []int) bool {

        total := 0
        count := 0

        for numeric != 0 {
            total += (numeric % 10) * multiples[len(multiples) - (1 + count)]

            count += 1

            numeric /= 10
        }

        return last == outputs[int(total % 11)]

}

func main () {
        fmt.Printf("Hello, Go!\n")

        fmt.Printf("true==%t\n", isNricValid("S8944027J"))
        fmt.Printf("false==%t\n", isNricValid("S89a4027J"))
        fmt.Printf("false==%t\n", isNricValid("123"))
        fmt.Printf("true==%t\n", isFinValid("G6075119Q"))
        fmt.Printf("false==%t\n", isFinValid("G60721a9Q"))
        fmt.Printf("false==%t\n", isFinValid("G619Q"))
}

